Implement functions for minimum-sample-size-calculators

achimbab opened the pull request:
Changelog category (leave one):
- New Feature

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
New functions minSampleSizeContinous and minSampleSizeConversion 


I Implemented two functions for minimum sample size calculators.

### `SELECT minSampleSizeContinous(20, 10, 0.05, 0.8, 0.05)`
  - Parameters
    - Mean of control-metric
    - Standard deviation of conrol-metric
    - Minimal Detectable Effect
    - Sufficient statistical power to detect a treatment effect
    - Significance level
  - Return
    - A tuple containing minimum_sample_size, detect_range_lower, and detect_range_upper.


### `SELECT minSampleSizeConversion(0.9, 0.01, 0.8, 0.05)`
  - Parameters
    - Mean of control-metric
    - Minimal Detectable Effect
    - Sufficient statistical power to detect a treatment effect
    - Significance level
  - Return
    - A tuple containing minimum_sample_size, detect_range_lower, and detect_range_upper.

These functions can be used in AB Test and others.

The descriptions of formulas are in [LINK](https://towardsdatascience.com/required-sample-size-for-a-b-testing-6f6608dd330a).

otheng03 committed:
Implement functions for minimum-sample-size-calculators

kitaisreal commented on src/Functions/minSampleSize.cpp:
```diff
@@ -0,0 +1,166 @@
+
+        DataTypePtr getReturnTypeImpl(const DataTypes & /*arguments*/) const override
+        {
+            return getReturnType();
```
Need to validate arguments types, and throw exception if arguments types mismatch.

achimbab replied:
I added codes in order to validate arguments.

kitaisreal commented on src/Functions/minSampleSize.cpp:
```diff
@@ -0,0 +1,166 @@
+
+        ColumnPtr executeImpl(const ColumnsWithTypeAndName & arguments, const DataTypePtr &, size_t input_rows_count) const override
+        {
+            MutableColumnPtr to{getReturnType()->createColumn()};
```
Current implementation is extremely slow.
First we can safely unwrap arguments to columns implementations, ColumnFloatUInt64, ColumnFloat64,  ColumnFloat64.
Second we can unwrap result tuple column in 3 columns UInt64, Float64, Float64, resize data to necessary rows count, and then just insert by index.

achimbab replied:
I unwrapped arguments and the result tuple.

otheng03 committed:
Validate arguments and unwrap arguments and the result to each columns.

otheng03 committed:
Have tests use roundBankers in 02206_minimum_sample_size

otheng03 committed:
Minor fix for tests

otheng03 committed:
Modify the sample_size type from UInt64 to Float64

otheng03 committed:
An integer type can be used for the baseline and sigma.

kitaisreal commented on src/Functions/minSampleSize.cpp:
```diff
@@ -0,0 +1,372 @@
+            const auto & arg_sigma = arguments[1];
+            ColumnPtr res;
+
+            const bool valid = castBothTypes(arg_baseline.type.get(), arg_sigma.type.get(), [&](const auto & left, const auto & right)
```
@achimbab not sure, why this is needed here ? It seems that we have fixed types for these functions, or not ?

achimbab replied:
@kitaisreal I removed this codes.

otheng03 committed:
Remove some codes

otheng03 committed:
Merge branch 'ClickHouse:master' into minSampleSize

kitaisreal commented on src/Functions/minSampleSize.cpp:
```diff
@@ -0,0 +1,247 @@
+            for (size_t row_num = 0; row_num < input_rows_count; ++row_num)
+            {
+                /// Mean of control-metric
+                Float64 baseline = col_baseline.getFloat64(row_num);
```
```c++
ColumnVector
Float64 getFloat64(size_t n) const override
{
    if constexpr (is_arithmetic_v<T>)
        return static_cast<Float64>(data[n]);
    else
        throw Exception(ErrorCodes::NOT_IMPLEMENTED, "Cannot get the value of {} as Float64", TypeName<T>);
}
```
it too heavy method, please use `getData` from Column to get array of Float64.

achimbab replied:
I fixed it to use `getData`.

kitaisreal commented on src/Functions/minSampleSize.cpp:
```diff
@@ -0,0 +1,247 @@
+
+                using namespace boost::math;
+                normal_distribution<> nd(0.0, 1.0);
+                Float64 min_sample_size = 2 * (std::pow(sigma, 2)) * std::pow(quantile(nd, 1.0 - alpha / 2) + quantile(nd, power), 2) / std::pow(delta, 2);
```
Please add original formulas and links into code.

achimbab replied:
I added the fomula and a link.

kitaisreal commented on src/Functions/minSampleSize.cpp:
```diff
@@ -0,0 +1,247 @@
+                Float64 delta = baseline * mde;
+
+                using namespace boost::math;
+                normal_distribution<> nd(0.0, 1.0);
```
It seems that distribution constructor should be moved out of hot loop.

achimbab replied:
I moved distribution constructor out of the hot loop.

kitaisreal commented on src/Functions/minSampleSize.cpp:
```diff
@@ -0,0 +1,247 @@
+                Float64 p_bar = (p1 + p2) / 2.0;
+
+                using namespace boost::math;
+                normal_distribution<> nd(0.0, 1.0);
```
Same as for `ContinousImpl`.

achimbab replied:
OK. I fixed it.

kitaisreal reviewed (approved):
Looks good. Please consider to make additional small improvements. Then it is ready to be merged.

kitaisreal reviewed (changes requested):
Need to fix small comments.

otheng03 committed:
Some fixes for minSampleSize functions.
- use getData
- add fomulas
- move distribution constructor out of the hot loop
- fix test cases
- use fixed const values for function arguments

otheng03 committed:
Merge branch 'minSampleSize' of https://github.com/achimbab/ClickHouse into minSampleSize

kitaisreal reviewed (approved):
Will merge in merging branch to update documentation and check performance.
